File: Android/app/src/main/java/com/cx/easyrtc/Activity/RTCActivity.java
package com.cx.easyrtc.Activity;

import android.content.Intent;
import android.opengl.GLSurfaceView;
import android.os.Bundle;
import android.util.Log;
import android.widget.ImageButton;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.cx.easyrtc.Agent.Agent;
import com.cx.easyrtc.EasyRTCApplication;
import com.cx.easyrtc.R;
import com.cx.easyrtc.Socket.SocketWraper;
import com.cx.easyrtc.WebRTC.WebRTCWraper;

import org.json.JSONException;
import org.json.JSONObject;
import org.webrtc.MediaStream;
import org.webrtc.VideoRenderer;
import org.webrtc.VideoRendererGui;

import java.util.ArrayList;

public class RTCActivity extends AppCompatActivity implements SocketWraper.SocketDelegate, WebRTCWraper.RtcListener{

//    private final String TAG = RTCActivity.class.getName();

    private String mCallStatus;

    private boolean mIfNeedAddStream;

    private WebRTCWraper mRtcWrapper;

    private VideoRenderer.Callbacks mRtcLocalRender;

    private VideoRenderer.Callbacks mRtcRemoteRender;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_rtc);

        Log.e("skruazzz", "RTCActivity onCreate");
        setUI();
        getCallStatus();
        getType();
        setGLView();
        setVideoRender();
    }

    @Override
    protected void onStop() {
        Log.e("skruazzz", "RTCActivity onStop");
        SocketWraper.shareContext().removeListener(this);

        super.onStop();
    }

    private void setUI() {
        setButton();
    }

    private void setButton() {
        ImageButton mCancelButton = findViewById(R.id.CancelButton);
        mCancelButton.setOnClickListener(view -> {
            Log.e("sliver", "cancel button clicked");
            try {
                SocketWraper.shareContext().emit("exit", "yes");
            } catch (JSONException e) {
                e.printStackTrace();
            }
            mRtcWrapper.exitSession();
            Intent intent = new Intent(RTCActivity.this, CallActivity.class);
            startActivity(intent);
        });
    }

    private void getCallStatus() {
        mCallStatus = getIntent().getExtras().getString("status");
    }

    private void getType() {
        String remoteType = getIntent().getExtras().getString("type");
        mIfNeedAddStream = !remoteType.equals("camera");
    }

    private void setGLView() {
        GLSurfaceView mGLView = findViewById(R.id.glview);
        mGLView.setPreserveEGLContextOnPause(true);
        mGLView.setKeepScreenOn(true);
        VideoRendererGui.setView(mGLView, this::setRtcWraper);
    }

    private void setVideoRender() {
        mRtcLocalRender = VideoRendererGui.create(0, 0, 100, 100,
                VideoRendererGui.ScalingType.SCALE_ASPECT_FILL, true);
        mRtcRemoteRender = VideoRendererGui.create(0, 0, 100,100,
                VideoRendererGui.ScalingType.SCALE_ASPECT_FILL, false);
    }

    private void setSocketWraper() {
        SocketWraper.shareContext().addListener(this);
    }

    private void setRtcWraper() {
        Log.e("sliver", "RTCActivity setRtcWraper");
        mRtcWrapper = new WebRTCWraper(this, VideoRendererGui.getEGLContext(), mIfNeedAddStream);
        setSocketWraper();
        createOfferOrAck();
    }

    private void createOfferOrAck() {
        if (mCallStatus.equals("send")) {
            mRtcWrapper.createOffer();
        } else if (mCallStatus.equals("recv")) {
            SocketWraper.shareContext().ack(true);
        }
    }

    private void processSignalMsg(String source, String target, String type, String value) {
        Log.e("sliver", "RTCActivity processSignalMsg " + type);
        if (target.equals(SocketWraper.shareContext().getUid())) {
            if (type.equals("offer")) {
                Log.e("sliver", "RTCActivity receive offer " + mRtcWrapper);
                mRtcWrapper.setRemoteSdp(type, value);
                mRtcWrapper.createAnswer();
            }

            if (type.equals("answer")) {
                mRtcWrapper.setRemoteSdp(type, value);
            }

            if (type.equals("exit")) {
                mRtcWrapper.exitSession();
                Intent intent = new Intent(RTCActivity.this, CallActivity.class);
                startActivity(intent);
            }
        } else {
            Log.e("sliver", "RTCActivity get error tag");
        }
    }

    @Override
    public void onUserAgentsUpdate(ArrayList<Agent> agents) {

    }

    @Override
    public void onDisConnect() {
        Log.e("sliver", "RTCActivity onDisConnect");
        runOnUiThread(() -> Toast.makeText(EasyRTCApplication.getContext(), "can't connect to server", Toast.LENGTH_LONG).show());
    }

    @Override
    public void onRemoteEventMsg(String source, String target, String type, String value) {
        processSignalMsg(source, target, type, value);
    }

    @Override
    public void onRemoteCandidate(int label, String mid, String candidate) {
        Log.e("sliver", "RTCActivity onRemoteCandidate");
        mRtcWrapper.setCandidate(label, mid, candidate);
    }

    @Override
    public void onLocalStream(MediaStream mediaStream) {
        Log.e("sliver", "RTCActivity onLocalStream");
        mediaStream.videoTracks.get(0).addRenderer(new VideoRenderer(mRtcLocalRender));
        VideoRendererGui.update(mRtcLocalRender, 75,75, 25, 25,
                VideoRendererGui.ScalingType.SCALE_ASPECT_FILL, true);
    }

    @Override
    public void onAddRemoteStream(MediaStream mediaStream) {
        Log.e("sliver", "RTCActivity onAddRemoteStream");
        mediaStream.videoTracks.get(0).addRenderer(new VideoRenderer(mRtcRemoteRender));
        VideoRendererGui.update(mRtcRemoteRender, 0, 0, 75, 75,
                VideoRendererGui.ScalingType.SCALE_ASPECT_FILL, true);
        if (mIfNeedAddStream) {
            VideoRendererGui.update(mRtcLocalRender, 75, 75, 25, 25,
                    VideoRendererGui.ScalingType.SCALE_ASPECT_FILL, true);
        }
    }

    @Override
    public void onRemoveRemoteStream(MediaStream mediaStream) {
        Log.e("sliver", "RTCActivity onRemoveRemoteStream");
        VideoRendererGui.update(mRtcLocalRender, 75, 75, 25, 25,
                VideoRendererGui.ScalingType.SCALE_ASPECT_FILL, true);
    }

    @Override
    public void onCreateOfferOrAnswer(String type, String sdp) {
        Log.e("sliver", "RTCActivity onCreateOfferOrAnswer");
        try {
            SocketWraper.shareContext().emit(type, sdp);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onIceCandidate(int label, String id, String candidate) {
        Log.e("sliver", "RTCActivity onIceCandidate");
        try {
            JSONObject msg = new JSONObject();
            msg.put("source", SocketWraper.shareContext().getUid());
            msg.put("target", SocketWraper.shareContext().getTarget());
            msg.put("label", label);
            msg.put("mid", id);
            msg.put("candidate", candidate);

            SocketWraper.shareContext().emit("candidate", msg);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }
}

File: Android/app/src/main/java/com/cx/easyrtc/activity/RTCActivity.kt
package com.cx.easyrtc.activity

import android.content.Context
import android.content.Intent
import android.media.AudioManager
import android.opengl.GLSurfaceView
import android.os.Bundle
import android.util.Log
import android.widget.ImageView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.cx.easyrtc.R
import com.cx.easyrtc.agent.Agent
import com.cx.easyrtc.socket.SocketWrapper
import com.cx.easyrtc.socket.SocketWrapper.SocketDelegate
import com.cx.easyrtc.webRTC.RTCAudioManager
import com.cx.easyrtc.webRTC.WebRTCWrapper
import com.cx.easyrtc.webRTC.WebRTCWrapper.RtcListener
import org.json.JSONException
import org.json.JSONObject
import org.webrtc.MediaStream
import org.webrtc.RendererCommon
import org.webrtc.VideoRenderer
import org.webrtc.VideoRendererGui
import java.util.*

class RTCActivity : AppCompatActivity(), SocketDelegate, RtcListener {

    companion object {
        private const val TAG = "RTCActivity"
    }

    private var mCallStatus: String? = null
    private var mIfNeedAddStream = false
    private var mRtcWrapper: WebRTCWrapper? = null
    private var mRtcLocalRender: VideoRenderer.Callbacks? = null
    private var mRtcRemoteRender: VideoRenderer.Callbacks? = null
    private val mAudioManager by lazy { RTCAudioManager.create(this) }

    private var isFrontCamera: Boolean = true
    private var isMicMute: Boolean = false
    private var isCameraOff: Boolean = false
    private var isSpeakerOff: Boolean = true

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_rtc)
        Log.e(TAG, "RTCActivity onCreate")
        setUI()
        callStatus
        type
        setGLView()
        setVideoRender()
        mAudioManager.selectAudioDevice(RTCAudioManager.AudioDevice.SPEAKER_PHONE)
        setSpeakerphoneOn()
    }

    override fun onStop() {
        Log.e(TAG, "RTCActivity onStop")
        SocketWrapper.shareContext().removeListener(this)
        super.onStop()
    }

    private fun setUI() {
        setButton()
    }

    private fun setButton() {
        val mCancelButton = findViewById<ImageView>(R.id.cancelButton)
        val mSwitchCamera = findViewById<ImageView>(R.id.switchButton)
        val mMuteMic = findViewById<ImageView>(R.id.micButton)
        val mCameraOff = findViewById<ImageView>(R.id.videoButton)
        val speakerButton = findViewById<ImageView>(R.id.speakerButton)

        mCancelButton.setOnClickListener {
            Log.e(TAG, "cancel button clicked")
            try {
                SocketWrapper.shareContext().emit("exit", "yes")
            } catch (e: JSONException) {
                e.printStackTrace()
            }
            mRtcWrapper!!.exitSession()
            val intent = Intent(this@RTCActivity, CallActivity::class.java)
            startActivity(intent)
        }

        mSwitchCamera.setOnClickListener {
            showToast("Switching camera")
            mRtcWrapper?.changeCamera(isFrontCamera)
            isFrontCamera = !isFrontCamera
        }

        mMuteMic.setOnClickListener {
            showToast(if (!isMicMute) "Unmuting mic" else "Muting mic")
            mRtcWrapper?.turnOffAudio(isMicMute)
            isMicMute = !isMicMute
        }

        mCameraOff.setOnClickListener {
            showToast(if (!isCameraOff) "Turning off your camera" else "Turning on your camera")
            mRtcWrapper?.turnOffCamera(isCameraOff)
            isCameraOff = !isCameraOff
        }

        speakerButton.setOnClickListener {
            showToast(if (!isSpeakerOff) "Speaker on" else "Speaker off")
            mRtcWrapper?.turnOffSpeaker(isSpeakerOff)
            isSpeakerOff = !isSpeakerOff
        }

    }

    private val callStatus: Unit
        get() {
            mCallStatus = intent.extras!!.getString("status")
        }
    private val type: Unit
        get() {
            val remoteType = intent.extras!!.getString("type")
            mIfNeedAddStream = remoteType != "camera"
        }

    private fun setGLView() {
        val mGLView = findViewById<GLSurfaceView>(R.id.glview)
        mGLView.preserveEGLContextOnPause = true
        mGLView.keepScreenOn = true
        VideoRendererGui.setView(mGLView) { setRtcWrapper() }
    }

    private fun setSpeakerphoneOn() {
        mAudioManager.setDefaultAudioDevice(RTCAudioManager.AudioDevice.SPEAKER_PHONE)
        val audioManager = getSystemService(Context.AUDIO_SERVICE) as AudioManager
        audioManager.isSpeakerphoneOn = true
        mRtcWrapper?.setSpeakerphoneOn(this)
    }

    private fun setVideoRender() {
        mRtcLocalRender = VideoRendererGui.create(
            0,
            0,
            100,
            100,
            RendererCommon.ScalingType.SCALE_ASPECT_FILL,
            true
        )

        mRtcRemoteRender = VideoRendererGui.create(
            0,
            0,
            100,
            100,
            RendererCommon.ScalingType.SCALE_ASPECT_FILL,
            false
        )
    }

    private fun setSocketWrapper() {
        SocketWrapper.shareContext().addListener(this)
    }

    private fun setRtcWrapper() {
        Log.e(TAG, "RTCActivity setRtcWrapper")
        mRtcWrapper = WebRTCWrapper(
            this,
            mIfNeedAddStream
        )
        setSocketWrapper()
        createOfferOrAck()
    }

    private fun createOfferOrAck() {
        if (mCallStatus == "send") {
            mRtcWrapper!!.createOffer()
        } else if (mCallStatus == "recv") {
            SocketWrapper.shareContext().ack(true)
        }
    }

    private fun processSignalMsg(target: String, type: String, value: String) {
        Log.e(TAG, "RTCActivity processSignalMsg $type")
        if (target == SocketWrapper.shareContext().uid) {
            if (type == "offer") {
                Log.e(TAG, "RTCActivity receive offer $mRtcWrapper")
                mRtcWrapper!!.setRemoteSdp(type, value)
                mRtcWrapper!!.createAnswer()
            }
            if (type == "answer") {
                mRtcWrapper!!.setRemoteSdp(type, value)
            }
            if (type == "exit") {
                mRtcWrapper!!.exitSession()
                val intent = Intent(this@RTCActivity, CallActivity::class.java)
                startActivity(intent)
            }
        } else {
            Log.e(TAG, "RTCActivity get error tag")
        }
    }

    override fun onUserAgentsUpdate(agents: ArrayList<Agent>) {}

    override fun onDisConnect() {
        Log.e(TAG, "RTCActivity onDisConnect")
        runOnUiThread {
            Toast.makeText(
                this,
                "can't connect to server",
                Toast.LENGTH_LONG
            ).show()
        }
    }

    override fun onRemoteEventMsg(source: String, target: String, type: String, value: String) {
        processSignalMsg(target, type, value)
    }

    override fun onRemoteCandidate(label: Int, mid: String, candidate: String) {
        Log.e(TAG, "RTCActivity onRemoteCandidate")
        mRtcWrapper!!.setCandidate(label, mid, candidate)
    }

    override fun onLocalStream(mediaStream: MediaStream) {
        Log.e(TAG, "RTCActivity onLocalStream")
        mediaStream.videoTracks[0].addRenderer(VideoRenderer(mRtcLocalRender))
        VideoRendererGui.update(
            mRtcLocalRender, 75, 75, 25, 25,
            RendererCommon.ScalingType.SCALE_ASPECT_FILL, true
        )
    }

    override fun onAddRemoteStream(mediaStream: MediaStream) {
        Log.e(TAG, "RTCActivity onAddRemoteStream")
        mediaStream.videoTracks[0].addRenderer(VideoRenderer(mRtcRemoteRender))
        VideoRendererGui.update(
            mRtcRemoteRender, 0, 0, 75, 75,
            RendererCommon.ScalingType.SCALE_ASPECT_FILL, true
        )
        if (mIfNeedAddStream) {
            VideoRendererGui.update(
                mRtcLocalRender, 75, 75, 25, 25,
                RendererCommon.ScalingType.SCALE_ASPECT_FILL, true
            )
        }
    }

    override fun onRemoveRemoteStream(mediaStream: MediaStream) {
        Log.e(TAG, "RTCActivity onRemoveRemoteStream")
        VideoRendererGui.update(
            mRtcLocalRender, 75, 75, 25, 25,
            RendererCommon.ScalingType.SCALE_ASPECT_FILL, true
        )
    }

    override fun onCreateOfferOrAnswer(type: String, sdp: String) {
        Log.e(TAG, "RTCActivity onCreateOfferOrAnswer")
        try {
            SocketWrapper.shareContext().emit(type, sdp)
        } catch (e: JSONException) {
            e.printStackTrace()
        }
    }

    override fun onIceCandidate(label: Int, id: String, candidate: String) {
        Log.e(TAG, "RTCActivity onIceCandidate")
        try {
            val msg = JSONObject()
            msg.put("source", SocketWrapper.shareContext().uid)
            msg.put("target", SocketWrapper.shareContext().target)
            msg.put("label", label)
            msg.put("mid", id)
            msg.put("candidate", candidate)
            SocketWrapper.shareContext().emit("candidate", msg)
        } catch (e: JSONException) {
            e.printStackTrace()
        }
    }

    private fun showToast(msg: String){
        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
    }
}